### PR TITLE
[Moco] Add support for MocoScaleFactor in other tracking goals

### PR DESCRIPTION
**Pull request by nickbianco**
Fixes issue #2903

### Brief summary of changes
Add MocoScaleFactor support for MocoStateTrackingGoal, MocoMarkerTrackingGoal, and MocoContactTrackingGoal.

### CHANGELOG.md (choose one)

- updated.

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022)

### CODE REVIEW

**Review by aymanhab — CHANGES_REQUESTED** [review]
> Reviewed 4 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022)* status: 4 of 9 files reviewed, 2 unresolved discussions (waiting on @carmichaelong and @nickbianco)

___
*[Bindings/Java/swig/java_moco.i, line 343 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeRdSwe9cVXhhrTPaJB:-MeRdSwe9cVXhhrTPaJC:b-ylyjdn) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/Bindings/Java/swig/java_moco.i#L343)):*
> ```SWIG
> %}
> 
> /* SWIG does not support initializer_list, but we can use Java arrays to
> ```

I'm assuming this was tested in Matlab. Do we have a mechanism to document this Java/Matlab specific changes that are not in Doxygen?
___
*[OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.h, line 173 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeRemz800uMxZYQUgE5:-MeRemz800uMxZYQUgE6:bttvvka) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.h#L173)):*
> ```objc
> Y = 1, Z = 2) of the scaled force value. The direction is applied in
> whatever frame the reference data is expressed in based on the provided
> ExternalLoads in each contact group.
> ```

At some point we were using positive-x, elsewhere in OpenSim we use X,-X, any reason to introduce index here instead?

**Review by aymanhab — COMMENTED** [review]
> Reviewed 4 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022)* status: 8 of 9 files reviewed, 2 unresolved discussions (waiting on @carmichaelong and @nickbianco)

**Review by carmichaelong — CHANGES_REQUESTED** [review]
Thanks @nickbianco. I was a little surprised that the `using RefPtrMSF...` statements needed to be in both the .h and .cpp files. Removing the line from the .cpp files still let it compile with MSVC, but I'd worry about it with other compilers (after reading https://stackoverflow.com/questions/57756881/where-to-put-using-directives-in-c-header-files). That being said, seems like it's working as intended and shouldn't pollute the namespace.

tl;dr I went down a rabbit hole but it's all good. Just some minor formatting comments.

> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-:-MeRgzqv8exaVrFkK6FM:b-gsufb6)* status: all files reviewed, 4 unresolved discussions (waiting on @nickbianco)

___
*[OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.cpp, line 84 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeQm8fKApt3jvzsmKFG:-MeQm8fKApt3jvzsmKFH:b7u6rjz) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.cpp#L84)):*
> ```cpp
>     for (int ig = 0; ig < getProperty_contact_groups().size(); ++ig) {
>         const auto& group = get_contact_groups(ig);
>         if (externalForceName ==  group.get_external_force_name()) {
> ```

Looks like there's an extra space here after `==`
___
*[OpenSim/Moco/MocoGoal/MocoMarkerTrackingGoal.h, line 133 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeRU8cy-trbJ7UMXKg5:-MeRU8cy-trbJ7UMXKg6:b-hlk6yn) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/OpenSim/Moco/MocoGoal/MocoMarkerTrackingGoal.h#L133)):*
> ```objc
>     mutable SimTK::Array_<std::string> m_marker_names;
>     mutable std::map<std::pair<std::string, int>, std::string>
>     m_scaleFactorMap;
> ```

add some indentation

**Review by nickbianco — COMMENTED** [review]
Thanks @carmichaelong. I honestly didn't even try not including it in the CPP file, I just assumed I'd need it there. Based on the SO page it looks like I need to make the alias `public`, but I think I'd prefer to keep it where it is. But like you said it's not an issue anyway.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-:-MeSQgiLBhuM32s3EAZ3:b-h8j3wd)* status: all files reviewed, 4 unresolved discussions (waiting on @aymanhab and @carmichaelong)

___
*[Bindings/Java/swig/java_moco.i, line 343 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeRdSwe9cVXhhrTPaJB:-MeSQ9Rh4A2nyGhxy5Be:b7hdc1o) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/Bindings/Java/swig/java_moco.i#L343)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

I'm assuming this was tested in Matlab. Do we have a mechanism to document this Java/Matlab specific changes that are not in Doxygen?
</blockquote></details>

The Matlab array type arguments will show up when you use `methodsview`, and we demonstrate the interface in a lot of the Moco examples. But we could have better documentation on Confluence for this stuff.
___
*[OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.h, line 173 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeRemz800uMxZYQUgE5:-MeSNZ_Q9v4AArp4sEQX:b-nbwhap) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.h#L173)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

At some point we were using positive-x, elsewhere in OpenSim we use X,-X, any reason to introduce index here instead?
</blockquote></details>

No particular reason, I was just trying to go with the option that made the most sense in each case. I've added a check to make sure the index is in the correct range.
___
*[OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.cpp, line 84 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeQm8fKApt3jvzsmKFG:-MeSNYU10ICMAcbWFdhx:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/OpenSim/Moco/MocoGoal/MocoContactTrackingGoal.cpp#L84)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

Looks like there's an extra space here after `==`
</blockquote></details>

Done.
___
*[OpenSim/Moco/MocoGoal/MocoMarkerTrackingGoal.h, line 133 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-MeRU8cy-trbJ7UMXKg5:-MeSNXoh9CqC1512iitR:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/9b05aca0caf7d17ac34676abcf7d3e1200ca3702/OpenSim/Moco/MocoGoal/MocoMarkerTrackingGoal.h#L133)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

add some indentation
</blockquote></details>

Done.

**Review by carmichaelong — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-:-MeVknlk-zFYAI8U3WVd:bnfp4nl)* status: 5 of 9 files reviewed, 2 unresolved discussions (waiting on @aymanhab and @carmichaelong)

**Review by nickbianco — COMMENTED** [review]
Any final comments @aymanhab?

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022#-:-MeVuOdR6ltT7sx5Qysn:b-2o2lbz)* status: 5 of 9 files reviewed, 2 unresolved discussions (waiting on @aymanhab and @carmichaelong)

**Review by aymanhab — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3022)* status: 5 of 9 files reviewed, all discussions resolved (waiting on @aymanhab and @carmichaelong)

**Comment by aymanhab** [comment]
LGTM, thanks for checking @nickbianco

**Comment by nickbianco** [comment]
Thanks @aymanhab and @carmichaelong!
